src/multiplex/client.rs

```diff
@@ -138,21 +241,138 @@ where
     }
 }
 
-// ===== ClientInner =====
+// ===== Pending =====
 
-struct Pending<Tag, Item> {
+/// A type used to track in-flight requests.
+///
+/// Each pending response contains an associated `Tag` that is provided
+/// from the `TagStore` used to uniquely identify a request/response pair.
+pub struct Pending<Tag, Response> {
     tag: Tag,
-    tx: tokio::sync::oneshot::Sender<ClientResponse<Item>>,
+    tx: tokio::sync::oneshot::Sender<ClientResponse<Response>>,
     span: tracing::Span,
 }
 
+impl<Tag, Response> Pending<Tag, Response> {
+    /// Get the tag associated with this pending request.
+    pub fn tag(&self) -> &Tag {
+        &self.tag
+    }
+}
+
+impl<Tag, Response> fmt::Debug for Pending<Tag, Response>
+where
+    Tag: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Pending")
+            .field("tag", &self.tag)
+            .field("span", &self.span)
+            .finish()
+    }
+}
+
+// ===== Builder =====
+
+/// The default service error handler.
+pub type DefaultOnServiceError<E> = Box<dyn FnOnce(E) + Send>;
+
+/// Builder for [`Client`] this is used to configure the transport, pending store
+/// and service_handler.
+///
+/// # Defaults
+///
+/// By default this builder only requires a transport and sets a default [`PendingStore`]
+/// and error handler. The default pending store is just a [`VecDeque`] and the default
+/// error handler is just an empty closure.
+pub struct Builder<
+    T,
+    E,
+    Request,
+    P = VecDequePendingStore<T, Request>,
+    F = DefaultOnServiceError<E>,
+> {
+    transport: T,
+    on_service_error: F,
+    pending_store: P,
+    _pd: PhantomData<fn(Request, E)>,
+}
+
+impl<T, E, Request, F, P> Builder<T, E, Request, P, F>
+where
+    T: Sink<Request> + TryStream + TagStore<Request, <T as TryStream>::Ok> + Send + 'static,
+    P: PendingStore<T, Request> + Send + 'static,
+    E: From<Error<T, Request>>,
+    E: 'static + Send,
+    Request: 'static + Send,
+    T::Ok: 'static + Send,
+    T::Tag: Send,
+    F: FnOnce(E) + Send + 'static,
+{
+    fn new(transport: T) -> Builder<T, E, Request, VecDequePendingStore<T, Request>> {
+        Builder {
+            transport,
+            on_service_error: Box::new(|_| {}),
+            pending_store: VecDequePendingStore::default(),
+            _pd: PhantomData,
+        }
+    }
+
+    /// Set the provided [`PendingStore`].
+    pub fn pending_store<P2>(self, pending_store: P2) -> Builder<T, E, Request, P2, F> {
+        Builder {
+            pending_store,
+            on_service_error: self.on_service_error,
+            transport: self.transport,
+            _pd: PhantomData,
+        }
+    }
+
+    /// Set the provided service error handler.
+    ///
+    /// If the `Client` errors, its error is passed to `on_service_error`.
```

Comment:
```suggestion
    /// If the [`Client`] encounters an error, it passes that error to `on_service_error` before exiting.
```

Shouldn't this then be called `on_client_error`?